File: services/nuxt-frontend/types/property.ts
export interface Coordinates{
    city: string
    state: string
    latitude: string
    longitude: string
}

export interface Capacity{
    max: number
    bedrooms: number
    beds: Float16Array
    bathrooms: number
}

export interface HouseRules{
    pets_allowed: boolean | null
    smoking_allowed: boolean | null
    events_allowed: boolean | null
} 
    
export interface Details{
    space_overview: string | undefined
    guest_access: string | undefined
    other_details: string | undefined
    neighborhood_description: string | undefined
    getting_around: string | undefined
}

export interface Property{
    id: string
    name: string
    picture_url: string
    coordinates: Coordinates
    amenities: string[]
    description: string
    summary: string
    capacity: Capacity
    house_rules: HouseRules
    details: Details
}
    
export interface Property_wTotal{
   property: Property
   total_before_taxes: string
}
    
export interface Review {
    name: string
    img_src: string
    date: string
    platform: string
    review_content: string
    rating: Float16Array
}

export interface SearchedProperty {
  uuid: string
  total_before_taxes: number
}

File: services/fastapi/app/models/property.py
from pydantic import BaseModel
from typing import List, Optional

# model for the address of a listing
class Coordinates(BaseModel):
    city: str
    state: str
    latitude: str
    longitude: str

# model for the capacity of a listing
class Capacity(BaseModel):
    max: int
    bedrooms: int
    beds: int
    bathrooms: float

# model for the rules of the house
class HouseRules(BaseModel): 
    pets_allowed: bool
    smoking_allowed: bool
    events_allowed: Optional[bool] = None

# model for details of the house
class Details(BaseModel):
    space_overview: str | None
    guest_access: str | None
    other_details: str | None
    neighborhood_description: str | None
    getting_around: str | None
    
# base listing model
class Property(BaseModel):
    id: str
    name: str
    picture_url: str
    coordinates: Coordinates
    amenities: List[str]
    description: str
    summary: str
    capacity: Capacity
    house_rules: HouseRules
    details: Details